File: day9/python/part1.py
# from collections import namedtuple
# from enum import Enum, auto


GROUP_START = '{'
GROUP_END = '}'
GARBAGE_START = '<'
GARBAGE_END = '>'

CANCEL = '!'


def process_stream(string):
    running_sum = 0
    curr_group_score = 0
    garbage_character_sum = 0
    ignore_char = False
    in_garbage = False
    for c in string:
        if ignore_char:
            ignore_char = False
        elif c is CANCEL:
            ignore_char = True
        elif c is GROUP_START and not in_garbage:
            curr_group_score += 1
        elif c is GROUP_END and not in_garbage:
            running_sum += curr_group_score
            curr_group_score -= 1
        elif c is GARBAGE_END and in_garbage:
            in_garbage = False
        elif c is GARBAGE_START and not in_garbage:
            in_garbage = True
        elif in_garbage:
            garbage_character_sum += 1
    return (running_sum, garbage_character_sum)


with open('../input.txt') as f:
    (group_score, garbage_characters) = process_stream(f.readline().strip())
    print('Part 1: Sum of all group is {}'.format(group_score))
    print('Part 2: Number of garbage_characters {}'.format(garbage_characters))

File: day9/ocaml/part1.ml
open Printf
open Core

type process_state = {
    total_score: int;
    group_score : int;
    garbage_character_sum : int;
    ignore_char : bool;
    in_garbage : bool
}

let process_char state curr_char =
    let {total_score; group_score; garbage_character_sum; ignore_char; in_garbage} = state in
    if ignore_char then {state with ignore_char=false}
    else match (curr_char, in_garbage) with
    | ('!', _) -> {state with ignore_char=true}
    | ('{', false) -> {state with group_score=(group_score + 1)}
    | ('}', false) -> {state with total_score=total_score + group_score; group_score=(group_score - 1)}
    | ('>', true) -> {state with in_garbage=false}
    | ('<', false) -> {state with in_garbage=true}
    | (_, true) -> {state with garbage_character_sum=(garbage_character_sum + 1)}
    | (_, false) -> state


let solve stream = 
    let init_state = {total_score=0; group_score=0; garbage_character_sum=0; ignore_char=false; in_garbage=false} in
    let processed = List.fold_left ~init:init_state ~f:process_char stream in
    (processed.total_score, processed.garbage_character_sum)

let explode str =
    let rec explode_recur i l =
        if i < 0 then l else explode_recur (i - 1) (str.[i] :: l) in
    explode_recur ((String.length str) - 1) []

let stream =
    In_channel.read_lines "../input.txt"
    |> List.hd
    |> Option.value_exn
    |> explode

let (part1, part2) = solve stream
let () = Printf.printf("Part 1: Group score is %d\n") part1
let () = Printf.printf("Part 2: Number of garbage characters are %d\n") part2
